lib/diamonds.ts

```diff
@@ -0,0 +1,137 @@
+import { Contract, constants } from 'ethers';
+
+// TODO: import IDiamond type
+
+export interface Facet {
+  target: string;
+  selectors: string[];
+}
+
+enum FacetCutAction {
+  Add,
+  Replace,
+  Remove,
+}
+
+export interface FacetCut extends Facet {
+  action: FacetCutAction;
+}
+
+export function getContractSignatures(contract: Contract) {
+  return Object.keys(contract.interface.functions);
+}
+
+export function getContractSelectors(contract: Contract) {
+  const signatures = getContractSignatures(contract);
+  const selectors = signatures.reduce((acc: string[], val: string) => {
+    acc.push(contract.interface.getSighash(val));
+    return acc;
+  }, []);
+
+  return selectors;
+}
+
+export function selectorExistsInFacet(selector: string, facets: Facet[]) {
+  for (const facet of facets) {
+    if (facet.selectors.includes(selector)) return true;
+  }
+
+  return false;
+}
+
+export function printFacetCuts(
+  target: string,
+  selectors: string[],
+  action: number = 0,
+) {
+  return {
+    target: target,
+    action: action,
+    selectors: selectors,
+  };
+}
+
+// adds unregistered selectors
+export async function addUnregisteredSelectors(
+  diamond: Contract,
+  contracts: Contract[],
+) {
+  const registeredFacets: Facet[] = await diamond.facets();
+
+  let facets: Facet[] = [];
+  contracts.forEach((contract) => {
+    facets.push({
+      target: contract.address,
+      selectors: getContractSelectors(contract),
+    });
+  });
+
+  let facetCuts: FacetCut[] = [];
+  // if selector is not found in the registered selectors then it should be
```

Comment:
This function can be duplicated, and modified a bit in order to add unregistered selectors, and replace registered selectors. 

Perhaps called `upgradeExistingAndAddSelectors`?

Reply:
There's definitely room to build a single function that will handle deployments and upgrades. I do have some concerns about being opinionated about how deployments and upgrades are done. From my conversations with devs who use EIP2535, I found that there is a wide range of preferences for doing this. For example, some people will replace all of the selectors when upgrading others will replace only the ones that are needed. The approach I took here was to make the least amount of changes when adding/ replacing/ removing selectors.